packages/go_router/lib/go_router.dart

```diff
@@ -6,24 +6,284 @@
+// ignore: prefer_mixin
+class GoRouter extends ChangeNotifier with NavigatorObserver {
+  /// Default constructor to configure a GoRouter with a routes builder
+  /// and an error page builder.
+  GoRouter({
+    required List<GoRoute> routes,
+    // TODO(johnpryan): Change to a route, improve error API
+    GoRouterPageBuilder? errorPageBuilder,
+    GoRouterWidgetBuilder? errorBuilder,
+    GoRouterRedirect? redirect,
+    Listenable? refreshListenable,
+    int redirectLimit = 5,
+    bool routerNeglect = false,
+    String? initialLocation,
+    // TODO(johnpryan): Deprecate this parameter
+    UrlPathStrategy? urlPathStrategy,
+    List<NavigatorObserver>? observers,
+    bool debugLogDiagnostics = false,
+    // TODO(johnpryan): Deprecate this parameter
+    GoRouterNavigatorBuilder? navigatorBuilder,
+    String? restorationScopeId,
+  }) {
+    if (urlPathStrategy != null) {
+      setUrlPathStrategy(urlPathStrategy);
+    }
+
+    setLogging(enabled: debugLogDiagnostics);
+    WidgetsFlutterBinding.ensureInitialized();
+
+    routeConfiguration = RouteConfiguration(
+      routes: routes,
+      topRedirect: redirect ?? (_) => null,
+      redirectLimit: redirectLimit,
+    );
+
+    routeConfiguration.validate();
+
+    routeInformationParser = GoRouterInformationParser(
+      configuration: routeConfiguration,
+      debugRequireGoRouteInformationProvider: true,
+    );
+    routeInformationProvider = GoRouteInformationProvider(
+        initialRouteInformation: RouteInformation(
+            location: _effectiveInitialLocation(initialLocation)),
+        refreshListenable: refreshListenable);
+
+    routerDelegate = GoRouterDelegate(
+      configuration: routeConfiguration,
+      errorPageBuilder: errorPageBuilder,
+      errorBuilder: errorBuilder,
+      routerNeglect: routerNeglect,
+      observers: <NavigatorObserver>[
+        ...observers ?? <NavigatorObserver>[],
+        this
+      ],
+      restorationScopeId: restorationScopeId,
+      // wrap the returned Navigator to enable GoRouter.of(context).go() et al,
+      // allowing the caller to wrap the navigator themselves
+      builderWithNav:
+          (BuildContext context, GoRouterState state, Navigator nav) =>
+              InheritedGoRouter(
+        goRouter: this,
+        child: navigatorBuilder?.call(context, state, nav) ?? nav,
+      ),
+    );
+    assert(() {
+      log.info('setting initial location $initialLocation');
+      return true;
+    }());
+  }
+
+  /// The route configuration for the app.
+  late final RouteConfiguration routeConfiguration;
+
+  /// The route information parser used by the go router.
+  late final GoRouterInformationParser routeInformationParser;
+
+  /// The router delegate used by the go router.
+  late final GoRouterDelegate routerDelegate;
+
+  /// The route information provider used by the go router.
+  late final GoRouteInformationProvider routeInformationProvider;
```

Comment:
This issue is interesting: https://github.com/flutter/flutter/issues/106863

It seems these are our options:

1. Do nothing. Users need to import files manually to use these fields 🤮 
2. Export more files. Make `GoRouterInformationParser`, `GoRouterDelegate`, and `GoRouteInformationProvider` public. This makes it harder for us to make breaking changes in the future.
3. Make this expose the "base" types that are already public (like `RouteInformationProvider` instead of `GoRouteInformationProvider`). This is a breaking change and it may not solve all users' scenarios. For example, a user might need something from `GoRouteInformationProvider` that isn't available from `RouteInformationProvider`.

What are your thoughts on this? This can be solved in a future PR, but I'm curious on what direction you have in mind.

Reply:
I left a [comment](https://github.com/flutter/flutter/issues/106863#issuecomment-1184739862) to see if there's a reason these need to be assigned to variables.

My first thought was to change the type from `GoRouteInformationParser` to `RouteInformationParser` but this produces a warning because we have the [always_specify_types](https://dart-lang.github.io/linter/lints/always_specify_types.html) lint enabled. So I think we should put `@visibleForTesting` on these classes and encourage users to use the methods defined on GoRouter instead. We could also add an `// ignore: always_specify_types` comment here.

Reply:
Answered it here: https://github.com/flutter/flutter/issues/106863#issuecomment-1185291219
Since I noticed this: if it would be a vote I would prefer option 3 over 2, but at the end fine with either one.

Reply:
I filed https://github.com/flutter/flutter/issues/107757, which I'd like to include in the next major version (5.0.0)